[inductor] Reinplace funcitonalized _scatter ops

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116899
* #116898
* __->__ #116789

This expands the reinplacing pass to also reinplace functionalized scatter ops
`select_scatter`, `slice_scatter` and `as_strided_scatter`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

peterbell10 committed:
[inductor] Reinplace funcitonalized _scatter ops

This expands the reinplacing pass to also reinplace functionalized scatter ops
`select_scatter`, `slice_scatter` and `as_strided_scatter`.

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -636,10 +636,92 @@ def remove_noop_ops(graph: torch.fx.Graph):
+        return self, src, dim, start, end, step
+
+    self_, src, dim, start, end, step = arg_extractor(*node.args, **node.kwargs)
+    with graph.inserting_before(node):
+        view = graph_call_function(
+            graph,
+            aten.slice.Tensor,
+            self_,
+            dim,
+            start,
+            end,
+            step,
+        )
+        graph_call_function(graph, aten.copy_.default, view, src)
```
If there are multiple `slice_scatter` on a tensor, would there still be unnecessary `copy_` in the graph? For the example below, the `reinplace_inplaceable_ops` will remove the epilogue `copy_` but not the intermediate one that this function adds? cc @leslie-fang-intel 
```
        view_13 = torch.ops.aten.reshape.default(view_as_real_1, [1, 32, 8, 64]);  view_as_real_1 = None
        slice_2 = torch.ops.aten.slice.Tensor(arg76_1, 0, 0, 1)
        slice_3 = torch.ops.aten.slice.Tensor(slice_2, 1, 1, 33)
        copy = torch.ops.aten.copy.default(slice_3, view_13);  slice_3 = view_13 = None
        slice_scatter = torch.ops.aten.slice_scatter.default(slice_2, copy, 1, 1, 33);  slice_2 = copy = None
        slice_scatter_1 = torch.ops.aten.slice_scatter.default(arg76_1, slice_scatter, 0, 0, 1);
        # epilogue
        copy_ = torch.ops.aten.copy_.default(arg76_1, slice_scatter_1);  arg76_1 = slice_scatter_1 = None
```

peterbell10 replied:
You're correct, we can only reinplace `slice_scatter_1` and intermediate buffer before the final copy cannot be removed since `slice_scatter` still depends on `arg76_1`. Did this example come from a model?

peterbell10 replied:
Okay, the python code to generate this is simply
```python
arg76_1[:1, :32].copy_(view_13)
```

That does seem bad. I'll see if I can do better with this.

jgong5 replied:
> You're correct, we can only reinplace `slice_scatter_1` and the intermediate buffer before the final copy cannot be removed since `slice_scatter` still depends on `arg76_1`. Did this example come from a model?

Yes, it comes from llama. @leslie-fang-intel added more context here: https://github.com/pytorch/pytorch/pull/116899#issuecomment-1880311080

peterbell10 committed:
Fix _generalized_scatter fake impl on "[inductor] Reinplace funcitonalized _scatter ops"


This expands the reinplacing pass to also reinplace functionalized scatter ops
`select_scatter`, `slice_scatter` and `as_strided_scatter`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented:
Closing as I've squashed this with the top PR in the stack
